fmpz/mul_2exp.c

```diff
@@ -9,38 +10,103 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= FLINT_BITS - 2)  /* Result fits inside a small fmpz */
+    {
+        if (COEFF_IS_MPZ(*f))
+            _fmpz_clear_mpz(*f);
+
+        *f = c1 << exp;
+    }
+    else if (c1bits <= FLINT_BITS - 2)   /* g is small */
+    {
+        ulong expred = exp % GMP_NUMB_BITS;
+        int alloc = 1 + (FLINT_BITS != GMP_NUMB_BITS) + exp / GMP_NUMB_BITS
+            + ((c1bits + expred) > GMP_NUMB_BITS);
+        mp_limb_t * limbs;
+        slong tmp;
+
+        /* Ensure enough limbs are allocated for f */
+        if (!COEFF_IS_MPZ(*f))
         {
-            fmpz_set_si(f, 0);
+            /* TODO: Initialize the new mpz with alloc limbs instead of
```

Comment:
I thought recent GMP did not do an allocation by default. So the todo here seems unnecessary.

Reply:
In `_fmpz_new_mpz` you find `mpz_init2`, which for GMP 6.2.1 contains `__GMP_ALLOCATE_FUNC_LIMBS` which translates into `__gmp_allocate_func` which has the default `__gmp_default_allocate` which calls `malloc`.

I just thought it would be more efficient to directly allocate X amounts of limbs rather than first allocating 2 limbs and then reallocating X limbs. Or perhaps the default allocation method is overwritten (I have no idea)?

Reply:
Ok, I was unaware they still did the 2 limb allocate anywhere.